File: fabui/python/boot.py
#!/usr/bin/env python
from serial_utils import SerialUtils
import ConfigParser, json, os, time, argparse
from subprocess import call

config = ConfigParser.ConfigParser()
config.read('/var/www/lib/config.ini')

open(config.get('task', 'lock_file'), 'w+').close()


parser = argparse.ArgumentParser()
parser.add_argument("-R", "--reset", action="store_true")
parser.add_argument("-f", "--flush", action="store_true")
parser.add_argument("-s", "--save",  action="store_true")
parser.add_argument("-d", "--debug", action="store_true")
args = parser.parse_args()

hardware_reset = args.reset
flush          = args.flush
save_settings  = args.save
debug          = args.debug

commands = []

if(os.path.isfile(config.get('printer', 'settings_file')) == False):
    call(["sudo php /var/www/fabui/index.php settings recreateSettingsFiles"],shell=True)
#read config steps/units
settings_file = open(config.get('printer', 'settings_file'))
settings = json.load(settings_file)

su = SerialUtils(debug = debug)

fw_version = su.fwVersion()
hw_version = su.hwVersion()


def saveSettings(data, file):
    ### save config file
    with open(file, 'w') as outfile:
        json.dump(data, outfile)
    outfile.close()

def customHardware(serialUtils, settings, settings_file):
    """
    Revision for customs edits
    """
    if(debug):
        print "Harware Revision: Custom"
    ### send extra commands
    with open(settings['custom']['overrides'], 'r') as custom_overrides_file:
        for line in custom_overrides_file:
            commands.append(line.strip())
    logic = 1 if settings['custom']['invert_x_endstop_logic'] == True else 0
    ### invert x endstop logic
    commands.append('M747 X{0}'.format(logic))
    #commands.append('M500')
    """ 
    save settings file
    """
    saveSettings(settings, settings_file)
    return None
def hardware1(serialUtils, settings, settings_file):
    """
    Rev1: September 2014 - May 2015
    - Original FABtotum
    """
    if(debug):
        print "Harware Revision: 1"
    ### save config file
    settings['hardware']['id'] = 1
    settings['feeder']['show'] = True
    settings['a'] = 177.777778
    saveSettings(settings, settings_file)
    return None

def hardware2(serialUtils, settings, settings_file):
    """
    Rev2: June 2015 - August 2015
    - Simplified Feeder (Removed the disengagement and engagement procedure), if you want you can update it easily following this Tutorial: Feeder update.
    - Bowden tube improvement (Added a protection external sleeve to avoid the bowden tube get stuck in the back panel).
    - Endstops logic inverted.
    """
    if(debug):
        print "Harware Revision: 2"
    ### invert x endstop logic
    commands.append('M747 X1')
    ### Maximum feedrates (mm/s):
    commands.append('M203 X550.00 Y550.00 Z15.00 E12.00')
    ### save config file
    settings['hardware']['id'] = 2
    settings['feeder']['show'] = True
    settings['a'] = 177.777778
    saveSettings(settings, settings_file)
    return None

def hardware3(serialUtils, settings, settings_file):
    if(debug):
        print "Harware Revision: 3"
    """
    Rev3: Aug 2015 - Jan 2016
    - Back panel modified to minimize bowden tube collisions
    - Hotplate V2 as standard duty hotplate
    - Reed sensor (Contactless sensor for the frontal door)
    - Head V1 (hybrid) discontinued
    - Milling Head V2 (store.fabtotum.com/eu/store/milling-head-v2.html).
    - Print head V2 (store.fabtotum.com/eu/store/printing-head-v2.html).
    """
    ### invert x endstop logic
    commands.append('M747 X1')
    ### Maximum feedrates (mm/s):
    commands.append('M203 X550.00 Y550.00 Z15.00 E12.00')
    ### save config file
    settings['hardware']['id'] = 3
    settings['feeder']['show'] = False
    settings['a'] = 177.777778
    saveSettings(settings, settings_file)
    return None

def hardware4(serialUtils, settings, settings_file):
    """
    Rev4:
    """
    if(debug):
        print "Harware Revision: 4"
    ### invert x endstop logic
    commands.append('M747 X1')
    ### Maximum feedrates (mm/s):
    ### save config file
    settings['hardware']['id'] = 4
    settings['feeder']['show'] = False
    settings['a'] = 88.888889
    saveSettings(settings, settings_file)
    return None

def hardware5(serialUtils, settings, settings_file):
    """
    - Rev5:
    """
    if(debug):
        print "Harware Revision: 5"
    ### invert x endstop logic
    commands.append('M747 X1')
    ### Maximum feedrates (mm/s):
    commands.append('M203 X550.00 Y550.00 Z15.00 E23.00')
    ### save config file
    settings['hardware']['id'] = 5
    settings['feeder']['show'] = False
    settings['a'] = 88.888889
    saveSettings(settings, settings_file)
    return None

def hybridHead(serialUtils, settings):
    if(debug):
        print "Set Hybrid Head"
    ### set pid
    commands.append('M301 P15 I5 D30')
    ### set head id 
    commands.append('M793 S1')
    return None
def printHead2(serialUtils, settings):
    if(debug):
        print "Set Printing Head"
    #commands = []
    ### set pid
    commands.append('M301 P20 I3.5 D30')
    ### set head id 
    commands.append('M793 S2')
    return None
def millHead2(serialUtils, settings):
    if(debug):
        print "Set Milling Head"
    ### set head id 
    commands.append('M793 S3')
    return None

def laserHead1(serialUtils, settings):
    if(debug):
        print "Set Laser Head"
    ### set head id 
    commands.append('M794 S4')

HW_VERSION_CMDS = {
 'custom' : customHardware,
 '1'      : hardware1,
 '2'      : hardware2,
 '3'      : hardware3,
 '4'      : hardware4,
 '5'      : hardware5
}

HEAD_VERSION_CMDS = {
 'hybrid'  : hybridHead,
 'print_v2': printHead2,
 'mill_v2' : millHead2,
 'laser_v1': laserHead1
}
"""
START BOOTSTRAP
"""
if(hardware_reset):
    if(debug):
        print "Hardware Reset"
    su.reset()
    time.sleep(3)

if(flush):
    if(debug):
        print "Flush"
    su.flush()

if(hardware_reset == False):
    if(debug):
        print "Not hardware Reset"
    commands.append('M300')

### rise probe
commands.append('M402')
### laser off
commands.append('M60 S0')
commands.append('M61 S0')
### set safety door open: enable/disable warnings
commands.append('M732 S{0}'.format(settings['safety']['door']))
### set collision-warning enable/disable warnings
commands.append('M734 S{0}'.format(settings['safety']['collision-warning']))
### set homing preferences
commands.append('M714 S{0}'.format(settings['switch']))

""" install head """
if(settings['hardware']['head']['type'] in HEAD_VERSION_CMDS):
    HEAD_VERSION_CMDS[settings['hardware']['head']['type']](su, settings)
else:
    hybridHead(su, settings)

""" exec hardware revisions """
if(settings['settings_type'] == 'custom'):
    customHardware(su, settings, config.get('printer', 'settings_file'))
elif hw_version in HW_VERSION_CMDS:
    HW_VERSION_CMDS[hw_version](su, settings, config.get('printer', 'settings_file'))
else:
    settings['feeder']['show'] = True
    saveSettings(settings, config.get('printer', 'settings_file'))
    
if(save_settings):    
    eeprom        = su.eeprom()
    settings['e'] = eeprom['steps_per_unit']['e']
    saveSettings(settings, config.get('printer', 'settings_file'))    
#### save all to EEPROM
commands.append('M500')
### alive machine
commands.append('M728')
### set ambient lights ###
commands.append('M701 S{0}'.format(settings['color']['r']))
commands.append('M702 S{0}'.format(settings['color']['g']))
commands.append('M703 S{0}'.format(settings['color']['b']))
su.sendGCode(commands)
su.close()
if os.path.isfile(config.get('task', 'lock_file')):
    os.remove(config.get('task', 'lock_file'))

File: fabui/script/boot.php
<?php
/**
 *  BOOT SCRIPT FILE - INITIALIZE MACHINE CUSTOM PARAMETERS
 * 
 */
require_once '/var/www/lib/config.php';
require_once '/var/www/lib/serial.php';
require_once '/var/www/fabui/application/config/production/fabtotum.php';
require_once '/var/www/lib/utilities.php';

$ini_array = parse_ini_file(SERIAL_INI);

//==== LOCK FILE
fopen(LOCK_FILE, "w");

define('TIME_TO_SLEEP', 0.3);

//==================================================================

if(!file_exists(FABUI_PATH.'config/config.json')){ //if config doesn't exists then create default
	create_default_config();
}

//load config
$json_config = json_decode(file_get_contents(FABUI_PATH.'config/config.json'), TRUE);

//init serial
$serial = new Serial;
$serial->deviceSet($ini_array['port']);
$serial->confBaudRate($ini_array['baud']);
$serial->confParity("none");
$serial->confCharacterLength(8);
$serial->confStopBits(1);
$serial->deviceOpen();
$serial -> serialflush();

//=== hw id ======================================================
$serial -> serialflush();
$serial->sendMessage('M763'.PHP_EOL);
sleep(TIME_TO_SLEEP);
$hw_id_reply = $serial->readPort(4096);
$hw_id = trim(str_replace('ok', '', $hw_id_reply));
//==================================================================
//rise probe
$serial -> serialflush();
$serial->sendMessage('M402'.PHP_EOL);
sleep(TIME_TO_SLEEP);
$alive_machine = $serial->readPort();
//alive machine
$serial -> serialflush();
$serial->sendMessage('M728'.PHP_EOL);
sleep(TIME_TO_SLEEP);
$alive_machine = $serial->readPort();


$serial -> serialflush();
$serial->sendMessage('M701 S'.$json_config['color']['r'].PHP_EOL); //red
$serial->sendMessage('M702 S'.$json_config['color']['g'].PHP_EOL); //green
$serial->sendMessage('M703 S'.$json_config['color']['b'].PHP_EOL); //blue
sleep(TIME_TO_SLEEP);
$colors = $serial->readPort();

//==================================================================
//set safety door open: enable/disable warnings
$serial -> serialflush();
$serial->sendMessage('M732 S'.$json_config['safety']['door'].PHP_EOL);
sleep(TIME_TO_SLEEP);
$safety_door = $serial->readPort();

//set collision-warning enable/disable warnings
$serial -> serialflush();
$serial->sendMessage('M734 S'.$json_config['safety']['collision-warning'].PHP_EOL);
sleep(TIME_TO_SLEEP);
$safety_door = $serial->readPort();

//==================================================================
//set homing preferences
$serial -> serialflush();
$serial->sendMessage('M714 S'.$json_config['switch'].PHP_EOL);
sleep(TIME_TO_SLEEP);
$homing_preferences = $serial->readPort();
//==================================================================

//==================================================================
//set head pids
$head = isset($json_config['hardware']['head']['type']) && $json_config['hardware']['head']['type'] != '' ? $json_config['hardware']['head']['type'] : 'hybrid';
$serial -> serialflush();
$serial->sendMessage($config['heads_pids'][$head].PHP_EOL);
sleep(0.1);
$serial -> sendMessage('M500' . PHP_EOL);
sleep(0.1);
$serial -> sendMessage('M793 S'.$config['heads_fw_id'][$head] . PHP_EOL);
sleep(0.1);
$serial -> sendMessage('M500' . PHP_EOL);
$serial -> deviceClose();
//==================================================================
$hw_id = (isset($json_config['settings_type']) && $json_config['settings_type'] == 'custom') ? 'custom' : $hw_id;
//include and exec specific hardware config settings
if(file_exists(dirname(__FILE__).'/hardware/settings_'.$hw_id.'.php')){
	include dirname(__FILE__).'/hardware/settings_'.$hw_id.'.php';
}else{
	
	$json_config = json_decode(file_get_contents(FABUI_PATH.'config/config.json'), TRUE);	
	$json_config['feeder']['show'] = true;
	file_put_contents(FABUI_PATH.'config/config.json', json_encode($json_config));
	
	
}
//==================================================================
$serial->deviceOpen();
$serial->sendMessage('M999'.PHP_EOL.'M728'.PHP_EOL);
$serial -> deviceClose();
shell_exec('sudo rm '.LOCK_FILE);
?>